crypto/proto.go

```diff
@@ -0,0 +1,293 @@
+	errInvalidMetaSig         = errors.New("invalid meta header's signature")
+	errMissingBodySig         = errors.New("missing body signature")
+	errInvalidBodySig         = errors.New("invalid body signature")
+	errNonOriginBodySig       = errors.New("body signature is set in non-origin verification header")
+)
+
+func newErrInvalidVerificationHeader(depth uint, cause error) error {
+	return fmt.Errorf("invalid verification header at depth %d: %w", depth, cause)
+}
+
+// SignedRequest is a generic interface of a signed NeoFS API request.
+type SignedRequest[BODY proto.Message] interface {
+	GetBody() BODY
+	GetMetaHeader() *session.RequestMetaHeader
+	GetVerifyHeader() *session.RequestVerificationHeader
+}
+
+// SignedResponse is a generic interface of a signed NeoFS API response.
+type SignedResponse[BODY proto.Message] interface {
+	GetBody() BODY
+	GetMetaHeader() *session.ResponseMetaHeader
+	GetVerifyHeader() *session.ResponseVerificationHeader
+}
+
+// SignRequestWithBuffer signs request parts using provided [neofscrypto.Signer]
+// according to the NeoFS API protocol, and returns resulting verification
+// header to attach to this request.
+//
+// Buffer is optional and free after the call.
+func SignRequestWithBuffer[BODY proto.Message](signer Signer, r SignedRequest[BODY], buf []byte) (*session.RequestVerificationHeader, error) {
+	var ln int
+	var err error
+	vhOriginal := r.GetVerifyHeader()
+
+	var bs []byte
+	signBody := vhOriginal == nil
+	if signBody { // body should be signed by the original sender only
+		buf, ln = encodeMessage(r.GetBody(), buf)
+		bs, err = signer.Sign(buf[:ln])
+		if err != nil {
+			return nil, fmt.Errorf("%w: %w", errSignBody, err)
+		}
+	}
+
+	buf, ln = encodeMessage(r.GetMetaHeader(), buf)
+	ms, err := signer.Sign(buf[:ln])
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errSignMeta, err)
+	}
+
+	buf, ln = encodeMessage(vhOriginal, buf)
+	vs, err := signer.Sign(buf[:ln])
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errSignVerifyOrigin, err)
+	}
+
+	scheme := refs.SignatureScheme(signer.Scheme())
+	pub := PublicKeyBytes(signer.Public())
+	res := &session.RequestVerificationHeader{
+		MetaSignature:   &refs.Signature{Key: pub, Sign: ms, Scheme: scheme},
+		OriginSignature: &refs.Signature{Key: pub, Sign: vs, Scheme: scheme},
+		Origin:          vhOriginal,
+	}
+	if signBody {
+		res.BodySignature = &refs.Signature{Key: pub, Sign: bs, Scheme: scheme}
+	}
+	return res, nil
+}
+
+// VerifyRequestWithBuffer checks whether verification header of the request is
+// formed according to the NeoFS API protocol.
+//
+// Buffer is optional and free after the call.
+func VerifyRequestWithBuffer[BODY proto.Message](r SignedRequest[BODY], buf []byte) error {
+	v := r.GetVerifyHeader()
+	if v == nil {
+		return errMissingVerifyHdr
+	}
+
+	b := r.GetBody()
+	m := r.GetMetaHeader()
+	bs := maxEncodedSize(b, m, v)
+	for m, v := m, v; ; {
```

Comment:
what `m, v := m, v` does?

Reply:
preserves `m, v`

Reply:
This can be read, but `mo, vo := m, v` might be a better choice with one line less code. But in general that's a recursive structure and iterations and recursion don't play well together, so a separate helper can make it much simpler overall.

Reply:
done